string.h

```diff
@@ -1,49 +1,73 @@
-//lang::CwC
+// lang::CwC
 #pragma once
 
 #include "object.h"
+#include <cstdio>
 #include <cstdlib>
 #include <cstring>
-#include <cstdio> 
 
 /**
- * An immutable String class representing a char*
- * author: chasebish */
+ * A String class representing a char[]
+ * author: chasebish
+ */
 class String : public Object {
 public:
-  /** CONSTRUCTORS & DESTRUCTORS **/
-
-  /* Creates a String copying s */
-  String(const char* s);
+  /**
+   * Creates a String by copying a string literal.
+   *
+   * @arg s string literal to be used as a source for this String object
+   */
+  String(const char *s);
 
-  /* Copies a String copying the value from s */
-  String(String* const s);
+  /**
+   * Copy constructor.
+   *
+   * @arg s source String object to be copied from
+   */
+  String(const String *s);
 
-  /* Clears String from memory */
   ~String();
 
+  /**
+   * Inherits the following from parent Object:
+   *
+   * bool equals(const Object *o) const;
+   * size_t hash() const;
+   */
 
-  /** INHERITED METHODS **/
-
-  /* Inherited from Object, generates a hash for a String */
-  size_t hash();
-
-  /* Inherited from Object, checks equality between an String and an Object */
-  bool equals(Object* const obj);
-
+  /**
+   * Compares strings based on alphabetical order according to the following
+   * ruleset:
+   *   < 0 -> this String is less than String s
+   *   = 0 -> this String is equal to String s
+   *   > 0 -> this String is greater than String s
+   *
+   * @arg s  the String to compare this against
+   * @return an int resulting from applying the above rules
+   */
+  int cmp(const String *s) const;
 
-  /** STRING METHODS **/
-  
-  /** Compares strings based on alphabetical order
-   * < 0 -> this String is less than String s
-   * = 0 -> this String is equal to String s
-   * > 0 -> this String is greater than String s
+  /**
+   * Concatenates the given String onto this String.
+   *
+   * @arg s  the source String to be concatenated onto this String
+   * @effect mutates this
    */
-  int cmp(String* const s);
+  void concat(const String *s);
 
-  /* Creates a new String by combining two existing Strings */
-  String* concat(String* const s);
+  /**
+   * Concatenates the given String onto this String, storing the result in
+   * dest.
+   *
+   * @arg s    the source string to be concatenated onto this
+   * @arg dest the destination for the string resulting from this concatenation
+   */
+  void concat(const String *s, String *dest) const;
```

Comment:
Returning allocated memory is a very bad idea and in my experience often leads to unintuitive memory leaks. Especially in a base class that we can assume will be allocated and used many times in a large scale `Map<String, Object>` based application.

The solution for this is usually the `dest` object idiom, and I'd strongly recommend that we use that for this sort of behavior. However, since we are already in the `String` class we are safe to interact directly with the internal representation of the data and mutate this String while still providing an immutable external interface.

I am happy to remove the mutating version of `concat`, but feel strongly that we should not use the first version which returns a `new String`.

Reply:
You've convinced me on the dest object over raw pointers :)
I'm indifferent to the mutable strings, I think.
I'm also thinking that, because the entire class is going to use this, maybe we rename the mutable concat to append or something if we want to keep it? Just for simplicity's sake - two names for two functions?

Reply:
Renamed `concat` to `append`, added `size` based constructor to allow for creating empty `dest` string.